src/Jackalope/Transport/DoctrineDBAL/Client.php

```diff
@@ -2084,27 +2082,36 @@ public function registerNodeTypes($types, $allowUpdate)
                 $query = "SELECT * FROM phpcr_type_nodes WHERE name = ?";
                 $result = $this->getConnection()->fetchColumn($query, array($type->getName()));
                 if ($result) {
-                    $this->getConnection()->delete('phpcr_type_nodes', array('node_type_id' => $result));
                     $this->getConnection()->delete('phpcr_type_props', array('node_type_id' => $result));
                     $this->getConnection()->delete('phpcr_type_childs', array('node_type_id' => $result));
                 }
             }
 
+            $data = array(
+                'name' => $type->getName(),
+                'supertypes' => implode(' ', $type->getDeclaredSuperTypeNames()),
+                'is_abstract' => $type->isAbstract() ? 1 : 0,
+                'is_mixin' => $type->isMixin() ? 1 : 0,
+                'queryable' => $type->isQueryable() ? 1 : 0,
+                'orderable_child_nodes' => $type->hasOrderableChildNodes() ? 1 : 0,
+                'primary_item' => $type->getPrimaryItemName(),
+            );
+
             try {
-                $this->getConnection()->insert('phpcr_type_nodes', array(
-                    'name' => $type->getName(),
-                    'supertypes' => implode(' ', $type->getDeclaredSuperTypeNames()),
-                    'is_abstract' => $type->isAbstract() ? 1 : 0,
-                    'is_mixin' => $type->isMixin() ? 1 : 0,
-                    'queryable' => $type->isQueryable() ? 1 : 0,
-                    'orderable_child_nodes' => $type->hasOrderableChildNodes() ? 1 : 0,
-                    'primary_item' => $type->getPrimaryItemName(),
-                ));
+                if ($allowUpdate && $result) {
+                    $this->getConnection()->update('phpcr_type_nodes', $data, array('node_type_id' => $result));
+                } else {
+                    $this->getConnection()->insert('phpcr_type_nodes', $data);
+                }
```

Comment:
What do you think about it now?

I don't like this because it queries database in cycle :(
![snimek obrazovky 2017-01-18 v 14 31 35](https://cloud.githubusercontent.com/assets/7081340/22065982/025ae788-dd8b-11e6-9701-4e8305e0ab76.png)